aten/src/ATen/native/BinaryOps.cpp

```diff
@@ -213,29 +215,22 @@ Tensor& true_divide_(Tensor& self, const Tensor& divisor) {
   return native::true_divide_out(self, self, divisor);
 }
 
+// NOTE: integer floor division != C++ integer division. Integer division in
+//   C++ is equivalent to torch.trunc(torch.true_divide(a, b)), while
+//   floor division is equivalent to torch.floor(torch.true_divide(a, b)).
+//   For example, in Python -5 // 2 = -3, while in C++ -5 / 2 = -2.
+// NOTE: floor divide, like Python and NumPy's floor divide, participates in
+//   type promotion.
 Tensor& floor_divide_out(Tensor& result, const Tensor& self, const Tensor& other) {
-  auto iter = TensorIterator::binary_op(result, self, other);
-  div_stub(iter.device_type(), iter);
-
-  if (result.is_floating_point()) {
-    result.trunc_();
-  }
-
+  const auto r = self.true_divide(other).floor_();
+  resize_output(result, r.sizes().vec());
+  result.copy_(r);
   return result;
 }
 
 Tensor floor_divide(const Tensor& self, const Tensor& other) {
-  Tensor result;
-  auto iter = TensorIterator::binary_op(result, self, other);
-
-  div_stub(iter.device_type(), iter);
-
-  auto out = iter.output();
-  if (out.is_floating_point()) {
-    out.trunc_();
-  }
-
-  return out;
+  const auto result_dtype = at::native::result_type(TensorList{self, other});
```

Comment:
why don't you call floor_divide_out?

Also, isn't true_divide guaranteed to give the same result type?

Reply:
We could call out.

true_divide will always return a float or complex type because it always promotes to float, but floor divide can return integer types because it participates in regular type promotion.